### ISSUE TITLE
Go: mismatch between json struct tags

### ISSUE THREAD

**Issue by doriandekoning** [golang/protobuf#1403]
**What version of protobuf and what language are you using?**
Version: acaef6a  (can be seen in the testprotos on master)

**What did you do?**
Generated proto files, used the grpc-gateway package which exposed a mismatch in the json fields.
However, the exact steps are not that relevant I believe since they can be seen in the testprotos in this repository, [here](https://github.com/protocolbuffers/protobuf-go/tree/acaef6adb7b0a1b0a0801768dd3baed9dc7470d4/internal/testprotos/conformance) for example (but if useful I could make a small example too).

**What did you expect to see?**

Either (I'm not sure what the official spec is):
```
OptionalInt32          int32                             `protobuf:"varint,1,opt,name=optional_int32,json=optionalInt32,proto3" json:"optionalInt32,omitempty"`
```

or:
```
OptionalInt32          int32                             `protobuf:"varint,1,opt,name=optional_int32,json=optional_int32,proto3" json:"optional_int32,omitempty"`
```

This is a problem because some code (grpc-gateway for example) uses the `protobuf` annotation field but the golang json package uses the `json` annotation. 

**What did you see instead?**

The tags regarding json in on the fields of the generated structs do not match:
```
	OptionalInt32          int32                             `protobuf:"varint,1,opt,name=optional_int32,json=optionalInt32,proto3" json:"optional_int32,omitempty"`
```

(Can be seen [here](https://github.com/protocolbuffers/protobuf-go/blob/acaef6adb7b0a1b0a0801768dd3baed9dc7470d4/internal/testprotos/conformance/test_messages_proto3.pb.go#L228))
**Anything else we should know about your project / environment?**

It can be seen in this repo :)

I raised this in grpc-go where I got pointed to this repository.
My apologies if I misunderstood and this is not the correct place to report this (suspected) issue (of if this is intended behaviour).

**Comment by puellanivis** [comment]
This has already been extensively discussed, and we cannot “fix” this without breaking other people’s code. See: Hyrum’s Law

Just one of the many issues that have been opened before: #1307

**Comment by doriandekoning** [comment]
Thanks for the quick reply, I missed the other issue in my quick search. Too bad that it cannot be changed/fixed easily. A bit disappointing to be honest, on one hand I understand (and respect) that great effort is put in not breaking anything for users relying on current behaviour, on the other hand there are the disclaimers in the readme regarding unspecified behaviour and bugs. 

The thing that is not really clear to me is why it would not be make a configurable option where the default value does not change any behaviour?

**Comment by puellanivis** [comment]
There is a “configurable option”: use `protojson` which correctly implements the the proto3→JSON mapping

Why not have a configurable option to fix the `json` tag itself? Because the `json` tag was a mistake. It was added in, because it seemed like it would be so easy to just add it, and viola. But then when a defined JSON mapping was defined, there was no way to compatibly and properly do the JSON mapping with the built in `encoding/json`, and so 🤦‍♀️ 😭 

This is also why the project is now hesitant to implement any “neat and easy expansion” that occasionally comes along. Sometimes, the idea is a cool idea, and it would be really neat to have… and then we think back on how we painted ourselves into a corner implementing JSON mapping early.
